add `query_name` as tag from Spark Structured Streaming metrics

otosky opened the pull request:
### What does this PR do?
Adds a `query_name` tag for Spark Structured Streaming metrics.  

Part of the implementation for Structured Streams added in #8078 included a regex pattern to capture and validate the `metric_name`s related to structured streaming, but didn't capture the stream's `query_name`, which is also a provided component within the metric key.

### Motivation
This is a necessary tag to have if a cluster is running more than 1 stream! Without the query_name tag there is no way to disambiguate which stream is producing the structured_streaming metrics for that cluster.

### Additional Notes
A query name is optionally set on the Structured Stream when [writing to a sink](https://spark.apache.org/docs/latest/structured-streaming-programming-guide.html#starting-streaming-queries).  If the queryName param is not supplied, you will see a UUID in it's place, as evidenced by the diff in `spark/tests/fixtures/metrics_json`.

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached

ofek reviewed (changes requested):
@otosky Thanks! Can we put this behind a flag in case of default high cardinality due to UUIDs?

otosky commented:
> @otosky Thanks! Can we put this behind a flag in case of default high cardinality due to UUIDs?

@ofek Definitely - is it preferred to have this default `False` and then have the client op-in in their config?

ofek commented:
Yes please!

sarah-witt commented:
Hi @otosky, when adding the config option, can you also include a warning in the description that the option should not be enabled unless the `queryName` param is supplied?

otosky commented:
@ofek & @sarah-witt 
added an `enable_query_name_tag` flag and I think I added documentation in the right places.  Let me know if I'm missing anything or if anything needs renaming.

sarah-witt commented on spark/datadog_checks/spark/spark.py:
```diff
@@ -642,11 +646,16 @@ def _spark_structured_streams_metrics(self, running_apps, addl_tags):
                         continue
                     groups = match.groupdict()
                     metric_name = groups['metric_name']
+                    query_name = groups.get("query_name", '')
```
```suggestion
```
Can we move this line to inside of the if statement?

sarah-witt commented on spark/datadog_checks/spark/spark.py:
```diff
@@ -642,11 +646,16 @@ def _spark_structured_streams_metrics(self, running_apps, addl_tags):
                     tags = ['app_name:%s' % str(app_name)]
                     tags.extend(addl_tags)
+
+                    if self._enable_query_name_tag:
+                        tags.append('query_name:%s' % str(query_name))
+
```
```suggestion
                    if self._enable_query_name_tag:
                        query_name = groups.get("query_name", '')
                        if query_name == '':
                             tags.append('query_name:%s' % str(query_name))
                        else:
                           ... log a debug message 

```
What do you think about this?

otosky replied:
Now that I'm looking more carefully at the regex pattern, I don't think it will even reach this code-path if there isn't a match for `query_name` in the full metric name, due to this check:

https://github.com/DataDog/integrations-core/blob/ff8747691daf68ef04e3789ddb50fa23a961e6c3/spark/datadog_checks/spark/spark.py#L643-L646

Would it suffice if I move the debug messages to those continuation blocks instead?
```python
for gauge_name, value in iteritems(response):
    match = STRUCTURED_STREAMS_METRICS_REGEX.match(gauge_name)
    if not match:
        self.log.debug("No regex match found for gauge: '%s'" % str(gauge_name))
        continue
    groups = match.groupdict()
    metric_name = groups['metric_name']
    if metric_name not in SPARK_STRUCTURED_STREAMING_METRICS:
        self.log.debug("Unknown metric_name encountered: '%s'" % str(metric_name))
        continue
    metric_name, submission_type = SPARK_STRUCTURED_STREAMING_METRICS[metric_name]
    tags = ['app_name:%s' % str(app_name)]
    tags.extend(addl_tags)

    if self._enable_query_name_tag:
        query_name = groups['query_name']
        tags.append('query_name:%s' % str(query_name))
 
    self._set_metric(metric_name, submission_type, value, tags=tags)
```

sarah-witt replied:
That looks good, thanks!

otosky replied:
Updated!

sarah-witt reviewed (commented):
Thanks for the updates! I have two more comments

apigirl reviewed (commented):
👍 looks good for docs

sarah-witt reviewed (approved):
Looks good, thanks!

otosky commented:
Awesome!  Is there anything I have to do to get that last "PR Labels" check green?

sarah-witt commented:
No, we can ignore that though!
